grr/server/grr_response_server/gui/api_plugins/osquery.py

```diff
@@ -0,0 +1,115 @@
+from grr_response_core.lib.rdfvalues import osquery as rdf_osquery
+
+from grr_response_proto.api import osquery_pb2 as api_osquery_pb2
+from grr_response_core.lib.rdfvalues import structs as rdf_structs
+
+from grr_response_server.gui import api_call_handler_base
+from grr_response_server.gui import api_call_context
+from grr_response_server import data_store
+
+from grr_response_server.gui.api_plugins import client as api_client
+from grr_response_server.gui.api_plugins import flow as api_flow
+
+from typing import Optional
+from typing import Text
+from typing import Generator
+
+
+class ApiGetOsqueryResultsArgs(rdf_structs.RDFProtoStruct):
+  """An RDF wrapper class for the arguments of Osquery exporter."""
+
+  protobuf = api_osquery_pb2.ApiGetOsqueryResultsArgs
+  rdf_deps = [
+      api_client.ApiClientId,
+      api_flow.ApiFlowId,
+  ]
+
+
+class ApiGetOsqueryResultsHandler(api_call_handler_base.ApiCallHandler):
+  """An API handler for the timeline exporter."""
+
+  args_type = ApiGetOsqueryResultsArgs
+
+  def Handle(
+      self,
+      args: ApiGetOsqueryResultsArgs,
+      context: Optional[api_call_context.ApiCallContext] = None,
+  ) -> api_call_handler_base.ApiBinaryStream:
+    """Handles requests for the exporting of Osquery flow results."""
+    client_id = str(args.client_id)
+    flow_id = str(args.flow_id)
+
+    flow_obj = data_store.REL_DB.ReadFlowObject(client_id, flow_id)
+    if flow_obj.flow_class_name != osquery.OsqueryFlow.__name__:
+      message = f"Flow '{flow_id}' is not an Osquery flow"
+      raise ValueError(message)
+
+    if args.format == ApiGetOsqueryResultsArgs.Format.CSV:  # pytype: disable=attribute-error
+      return _StreamCsv(client_id=client_id, flow_id=flow_id)
+
+    message = f"Incorrect Osquery results export format: {args.format}"
+    raise ValueError(message)
+
+
+def _StreamCsv(
+    client_id: Text,
+    flow_id: Text,
+) -> api_call_handler_base.ApiBinaryStream:
+  filename = f"osquery_{flow_id}.csv"
+
+  results = _FetchOsqueryResults(client_id=client_id, flow_id=flow_id)
+  content = _ParseToCsvBytes(results)
+
+  return api_call_handler_base.ApiBinaryStream(filename, content)
+
+
+def _FetchOsqueryResults(
+    client_id: Text, 
+    flow_id: Text
+) -> Generator[rdf_osquery.OsqueryResult, None, None]:
+  next_to_fetch = 0
+  last_fetched_count = None
+
+  while last_fetched_count != 0:
+    data_fetched = data_store.REL_DB.ReadFlowResults(
+      offset=next_to_fetch,
+      count=_RESULTS_TO_FETCH_AT_ONCE,
+      client_id=client_id,
+      flow_id=flow_id)
+
+    last_fetched_count = len(data_fetched)
+    next_to_fetch += last_fetched_count
+
+    for datum in data_fetched:
+      if type(datum.payload) != rdf_osquery.OsqueryResult:
+        raise ValueError(f"Incorrect payload type: {type(datum.payload)}")
+      yield datum.payload
+
+
+def _ParseToCsvBytes(
+    osquery_results: Generator[rdf_osquery.OsqueryResult, None, None]
+) -> Generator[bytes, None, None]:
```

Comment:
`Iterator[bytes]`

Reply:
`Iterator` is a super-type of `Generator` so `Iterator[bytes]` is technically correct, however this function actually returns a `Generator` from my experiments (`<generator object _ParseToCsvBytes at 0x7f89525b8740>`), and also the [Python docs](https://docs.python.org/3.6/glossary.html#term-generator).

Reply:
Okay I agree that `Iterator[bytes]` is the right choice here. To answer my own question - yes the exact type is `Generator[bytes, None, None]`, however this is just an implementation detail; the intended exposed interface only needs an `Iterator` to iterate over.